src/GraphQL/Resolvers/SourceStreamMethodResolver.cs

```diff
@@ -86,16 +109,48 @@ public SourceStreamMethodResolver(MethodInfo methodInfo, LambdaExpression instan
                         bodyExpression);
                 }
             }
+            // IAsyncEnumerable<T>
+            else if (bodyExpression.Type.IsGenericType && bodyExpression.Type.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>))
+            {
+                var innerType = bodyExpression.Type.GetGenericArguments()[0];
+                var method = _convertFromAsyncEnumerableMethodInfo.MakeGenericMethod(innerType);
+                var func = method.CreateDelegate<Func<Expression, ParameterExpression, Func<IResolveFieldContext, ValueTask<IObservable<object?>>>>>();
+                return func(bodyExpression, resolveFieldContextParameter);
+            }
 
             if (taskBodyExpression == null)
             {
-                throw new InvalidOperationException("Method must return a IObservable<T> or Task<IObservable<T>> where T is a reference type.");
+                throw new InvalidOperationException("Method must return a IObservable<T> or Task<IObservable<T>> where T is a reference type, or a IAsyncEnumerable<T> or Task<IAsyncEnumerable<T>>.");
```

Comment:
```suggestion
                throw new InvalidOperationException("Method must return either: 1) a IObservable<T> or IAsyncEnumerable<T> or 2) Task<> or ValueTask<> containing types from 1. In all cases T must be a reference type.");
```

Something like that?

Reply:
Well, I'd say it looks a bit confusing.  And `IAsyncEnumerable<T>` does not require `T` to be a reference type.

Reply:
Where reference constraint come from?

Reply:
`IObservable<T>` is cast directly to `IObservable<object>`.  This kind of cast is only possible with reference types.  `IAsyncEnumerable<T>` performs a cast (including boxing if necessary) within the code that converts it directly to `IObservable<object>` -- specifically on ObservableFromAsyncEnumerable.cs:70 -- `observer.OnNext(z)` performs an implicit cast from T to object even if T is a value type.

Reply:
Note that `ObservableFromAsyncEnumerable<T>` implements `IObservable<object?>`, not `IObservable<T>`.  This was written this way because GraphQL.NET requires `IObservable<object?>` and so I put the cast directly in the implementation which allows value types to be returned.

Reply:
Ah, I see.

Reply:
I could write a wrapper for `IObservable<T>` value types so they are supported also.

Reply:
```c#
 private sealed class ObservableAdapter<T> : IObservable<object?>
        {
            private readonly IObservable<T> _observable;

            public ObservableAdapter(IObservable<T> observable)
            {
                _observable = observable;
            }

            public IDisposable Subscribe(IObserver<object?> observer) => _observable.Subscribe(new ObserverAdapter(observer));

            private sealed class ObserverAdapter : IObserver<T>
            {
                private readonly IObserver<object?> _observer;
                public ObserverAdapter(IObserver<object?> observer)
                {
                    _observer = observer;
                }
                public void OnCompleted() => _observer.OnCompleted();
                public void OnError(Exception error) => _observer.OnError(error);
                public void OnNext(T value) => _observer.OnNext(value);
            }
        }
```

Reply:
```c#
 private static async ValueTask<IObservable<object?>> CastFromValueTaskAsync<T>(ValueTask<IObservable<T>> task) where T : class
            => await task.ConfigureAwait(false);
        private static async ValueTask<IObservable<object?>> CastFromValueTask2Async<T>(ValueTask<IObservable<T>> task)
          => new ObservableAdapter<T>(await task.ConfigureAwait(false));
```

Reply:
> But I'm not sure that's much better.

Agree. With removal of T constraint it will be even more better.

Reply:
Thanks.  I'll do this is a subsequent PR.  It won't be hard but it will change enough that it will be easier to review separately.

Reply:
OK

Reply:
Should we somehow encourage to use ref-types instead of value-types for observables in error messages/docs ?

Reply:
No.  100% of GraphQL.NET boxes value types when returned from `IFieldResolver.ResolveAsync` and for storage within `ExecutionResult`.  Adding a wrapper here will not contribute any significant memory or CPU usage to processing time; it's a single allocation when subscribing only.  I also cannot find anything within the GraphQL spec that says that a scalar cannot be returned from a subscription rather than an object type.